test/xcdr/xcdrv1.cpp

```diff
@@ -0,0 +1,289 @@
+    size_t total_size = expected_streams[tested_stream].size() + 4; // Extra needed space for temp LongMemberHeader
+    auto buffer =
+            std::unique_ptr<char, void (*)(
+        void*)>{reinterpret_cast<char*>(calloc(total_size, sizeof(char))), free};
+    FastBuffer fast_buffer(buffer.get(), total_size);
+    Cdr cdr(fast_buffer, endianness, CdrVersion::XCDRv1);
+    //}
+
+    //{ Encode optional not present.
+    cdr.set_encoding_flag(encoding);
+    cdr.serialize_encapsulation();
+    Cdr::state enc_state(cdr);
+    cdr.begin_serialize_type(enc_state, encoding);
+    cdr.serialize_member(MemberId(1), value, header_selection);
+    cdr.end_serialize_type(enc_state);
+    Cdr::state enc_state_end(cdr);
+    //}
+
+    //{ Test encoded content
+    ASSERT_EQ(cdr.get_serialized_data_length(), expected_streams[tested_stream].size());
+    ASSERT_EQ(0, memcmp(buffer.get(), expected_streams[tested_stream].data(), expected_streams[tested_stream].size()));
+    //}
+
+    //{ Decoding optional not present
+    _T dec_value {0};
+    cdr.reset();
+    cdr.read_encapsulation();
+    ASSERT_EQ(cdr.get_encoding_flag(), encoding);
+    ASSERT_EQ(cdr.endianness(), endianness);
+    cdr.deserialize_type(encoding, [&dec_value](Cdr& cdr_inner, const MemberId& mid) -> bool
+            {
+                bool ret_value = true;
+                switch (mid.id)
+                {
+                    case 1:
+                        cdr_inner.deserialize_member(dec_value);
+                        break;
+                    default:
+                        ret_value = false;
+                        break;
+                }
+
+                return ret_value;
+            });
+    ASSERT_EQ(value, dec_value);
+    Cdr::state dec_state_end(cdr);
+    ASSERT_EQ(enc_state_end, dec_state_end);
+    //}
+}
+
+/*!
+ * @test Test the decoder can get the XCDR version from the encapsulation.
+ * @code{.idl}
+ * struct AutoSelectStruct
+ * {
+ *     @id(0)
+ *     unsigned short value1;
+ *     @id(1)
+ *     unsigned long value2;
+ * };
+ * @endcode
+ */
+TEST_P(XCdrv1Test, auto_selection_on_decode)
+{
+    EncodingAlgorithmFlag encoding = std::get<0>(GetParam());
+    Cdr::Endianness endianness = Cdr::Endianness::LITTLE_ENDIANNESS;
+    auto buffer =
+            std::unique_ptr<char, void (*)(
+        void*)>{reinterpret_cast<char*>(calloc(100, sizeof(char))), free};
+    FastBuffer fast_buffer(buffer.get(), 100);
+
+    //{ Encode a ushort and a ulong.
+    Cdr cdr(fast_buffer, endianness, CdrVersion::XCDRv1);
+    const uint16_t us {0x01FC};
+    const uint32_t ul {0x01FC1FCD};
+    cdr.set_encoding_flag(encoding);
+    cdr.serialize_encapsulation();
+    Cdr::state enc_state(cdr);
+    cdr.begin_serialize_type(enc_state, encoding);
+    cdr << MemberId(0) << us;
+    cdr << MemberId(1) << ul;
+    cdr.end_serialize_type(enc_state);
+    //}
+
+    //{ Decode a ushort and a ulong.
```

Comment:
Typo:

```suggestion
    //{ Decode an ushort and an ulong.
```

Please, fix also other occurrences in PR

Reply:
Done in 07380dc

Reply:
There are some more occurrences of this suggestion in this file and in xcdrv2.cpp file.

Reply:
Done in c459712